samcli/local/lambda_service/local_lambda_invoke_service.py

```diff
@@ -115,47 +118,113 @@ def _construct_error_handling(self):
 
         Parameters
         ----------
         function_name str
-            Name of the function to invoke
+            Name of function to invoke
+
+        request_data str
+            Parameters to pass to lambda function
+
+        stdout io.BaseIO
+            Output stream that stdout should be written to
+
+        stderr io.BaseIO
+            Output stream that stderr should be written to
 
         Returns
         -------
-        A Flask Response response object as if it was returned from Lambda
+        A flask Response response object as if it was returned from Lambda
         """
-        flask_request = request
+        request_id = uuid4()
 
-        request_data = flask_request.get_data()
+        thread = threading.Thread(target=self._invoke_sync_request,
+                                  args=(function_name, request_data),
+                                  kwargs={'stdout': stdout, 'stderr': stderr})
+        thread.daemon = True
+        thread.start()
 
-        if not request_data:
-            request_data = b'{}'
+        LOG.debug('Async invocation: %s, requestId=%s', function_name, request_id)
+        return self.service_response(None,
+                                     {'x-amzn-requestid': request_id},
+                                     202)
 
-        request_data = request_data.decode('utf-8')
+    def _invoke_sync_request(self, function_name, request_data, stdout, stderr):
+        """
+        Synchronous Request Handler for the Local Lambda Invoke path.
 
-        stdout_stream = io.BytesIO()
+        Parameters
+        ----------
+        function_name str
+            Name of function to invoke
+
+        request_data str
+            Parameters to pass to lambda function
+
+        stdout io.BaseIO
+            Output stream that stdout should be written to
+
+        stderr io.BaseIO
+            Output stream that stderr should be written to
+
+        Returns
+        -------
+        A flask Response response object as if it was returned from Lambda
+        """
 
         try:
-            self.lambda_runner.invoke(function_name, request_data, stdout=stdout_stream, stderr=self.stderr)
+            self.lambda_runner.invoke(function_name, request_data, stdout=stdout, stderr=stderr)
         except FunctionNotFound:
             LOG.debug('%s was not found to invoke.', function_name)
             return LambdaErrorResponses.resource_not_found(function_name)
 
         lambda_response, lambda_logs, is_lambda_user_error_response = \
-            LambdaOutputParser.get_lambda_output(stdout_stream)
+            LambdaOutputParser.get_lambda_output(stdout)
 
-        if self.stderr and lambda_logs:
+        if stderr and lambda_logs:
             # Write the logs to stderr if available.
-            self.stderr.write(lambda_logs)
+            stderr.write(lambda_logs)
 
         if is_lambda_user_error_response:
+            LOG.debug('Lambda error response: %s', lambda_response)
```

Comment:
Is this really needed if we are sending this data back in the next line? My opinion is to remove this.